# Readmission Classification - Data Cleaning

[Jose R. Zapata](https://joserzapata.github.io)
- https://joserzapata.github.io
- https://twitter.com/joserzapata
- https://www.linkedin.com/in/jose-ricardo-zapata-gonzalez/       


<h1>Contenido<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Readmission-Classification---Data-Cleaning" data-toc-modified-id="Readmission-Classification---Data-Cleaning-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Readmission Classification - Data Cleaning</a></span><ul class="toc-item"><li><span><a href="#Introduction" data-toc-modified-id="Introduction-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Introduction</a></span></li><li><span><a href="#Main-Objective" data-toc-modified-id="Main-Objective-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Main Objective</a></span></li><li><span><a href="#Principal-References" data-toc-modified-id="Principal-References-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Principal References</a></span><ul class="toc-item"><li><span><a href="#Paper" data-toc-modified-id="Paper-1.3.1"><span class="toc-item-num">1.3.1&nbsp;&nbsp;</span>Paper</a></span></li><li><span><a href="#Dataset" data-toc-modified-id="Dataset-1.3.2"><span class="toc-item-num">1.3.2&nbsp;&nbsp;</span>Dataset</a></span></li><li><span><a href="#Data-description" data-toc-modified-id="Data-description-1.3.3"><span class="toc-item-num">1.3.3&nbsp;&nbsp;</span>Data description</a></span></li></ul></li><li><span><a href="#Paper-Insigths" data-toc-modified-id="Paper-Insigths-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Paper Insigths</a></span></li></ul></li><li><span><a href="#Import-libraries" data-toc-modified-id="Import-libraries-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Import libraries</a></span></li><li><span><a href="#Load-Datasets" data-toc-modified-id="Load-Datasets-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Load Datasets</a></span></li><li><span><a href="#Data-Description-and-Cleaning" data-toc-modified-id="Data-Description-and-Cleaning-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Data Description and Cleaning</a></span><ul class="toc-item"><li><span><a href="#Remove-duplicated-and-Not-necesary-Information" data-toc-modified-id="Remove-duplicated-and-Not-necesary-Information-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Remove duplicated and Not necesary Information</a></span></li><li><span><a href="#Handling-Missing-Values" data-toc-modified-id="Handling-Missing-Values-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Handling Missing Values</a></span><ul class="toc-item"><li><span><a href="#Removing-missing-values" data-toc-modified-id="Removing-missing-values-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>Removing missing values</a></span><ul class="toc-item"><li><span><a href="#race-column-analysis" data-toc-modified-id="race-column-analysis-4.2.1.1"><span class="toc-item-num">4.2.1.1&nbsp;&nbsp;</span>race column analysis</a></span></li></ul></li><li><span><a href="#Save-intermediated-transformed-data" data-toc-modified-id="Save-intermediated-transformed-data-4.2.2"><span class="toc-item-num">4.2.2&nbsp;&nbsp;</span>Save intermediated transformed data</a></span></li></ul></li></ul></li><li><span><a href="#Partial-Results" data-toc-modified-id="Partial-Results-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Partial Results</a></span></li><li><span><a href="#References" data-toc-modified-id="References-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>References</a></span></li></ul></div>

## Introduction


Analyze factors related to readmission as well as other outcomes pertaining to patients in order to classify a patient-hospital outcome

3 different outputs:

1. No readmission

2. A readmission in less than `30` days (this situation is not good, because maybe your treatment was not appropriate);

3. A readmission in more than 30 days (this one is not so good as well the last one, however, the reason could be the state of the patient.


## Main Objective

> **How effective was the treatment received in hospital?**

## Principal References

### Paper

Beata Strack, Jonathan P. DeShazo, Chris Gennings, Juan L. Olmo, Sebastian Ventura, Krzysztof J. Cios, and John N. Clore, “Impact of HbA1c Measurement on Hospital Readmission Rates: Analysis of 70,000 Clinical Database Patient Records,” BioMed Research International, vol. 2014, Article ID 781670, 11 pages, 2014.

https://www.hindawi.com/journals/bmri/2014/781670/

### Dataset

https://archive.ics.uci.edu/ml/datasets/Diabetes+130-US+hospitals+for+years+1999-2008#

### Data description

https://www.hindawi.com/journals/bmri/2014/781670/tab1/

## Paper Insigths

Multivariable logistic regression was used to fit the relationship between the measurement of HbA1c and **early readmission** while controlling for covariates such as demographics, severity and type of the disease, and type of admission.

The measurement of HbA1c was performed infrequently (18.4%) in the inpatient setting.

The statistical model suggests that the relationship between the probability of readmission and the HbA1c measurement depends on the primary diagnosis.

The data suggest further that the greater attention to diabetes reflected in HbA1c determination may improve patient outcomes and lower cost of inpatient care.

Databases of clinical data contain valuable but heterogeneous and difficult data in terms of missing values, incomplete or inconsistent records, and high dimensionality understood not only by number of features but also their complexity.

preliminary analysis and preprocessing of the data were performed resulting in retaining only these features (attributes) and encounters that could be used in further analysis, that is, contain sufficient information.

** criteria for extracting the data**

1.	It is an inpatient encounter (a hospital admission).
2.	It is a “diabetic” encounter, that is, one during which any kind of diabetes was entered to the system as a diagnosis.
3.	The length of stay was at least 1 day and at most 14 days.
4.	Laboratory tests were performed during the encounter.
5.	Medications were administered during the encounter.

Criteria  were applied to remove admissions for procedures and so forth, which were of less than 23 hours of duration and in which changes in diabetes management were less likely to have occurred. 

It should be noted that the diabetic encounters are not all encounters of diabetic patients but rather only these encounters where diabetes was coded as an existing health condition.

primarily interested in factors that lead to early readmission

- “readmitted,” if the patient was readmitted within 30 days of discharge
- “otherwise,” which covers both readmission after 30 days and no readmission at all.

Note that 30 days was chosen based on criteria often used by funding agencies

Hemoglobin A1c (HbA1c) is an important measure of glucose control, which is widely applied to measure performance of diabetes care.

The database contains detailed information about dosage but is restricted only to medications administered during the encounter. It was not possible to track any preadmission and discharge medications.

They considered four groups of encounters: (1) no HbA1c test performed, (2) HbA1c performed and in normal range, (3) HbA1c performed and the result is greater than 8% with no change in diabetic medications, and (4) HbA1c performed, result is greater than 8%, and diabetic medication was changed.

**Data information**

the original database contains incomplete, redundant, and noisy information as expected in any real-world data. weight (97% values missing), payer code (40%), and medical specialty (47%). Weight attribute was not included in further analysis. Payer code was removed. Medical specialty attribute was maintained.

The preliminary dataset contained multiple inpatient visits for some patients and the observations could **not be considered as statistically independent**

removed all encounters that resulted in either **discharge to a hospice** or **patient death**, to avoid biasing our analysis.

Processed dataset consists of hospital admissions of length between one and 14 days that did not result in a patient death or discharge to a hospice.

**Each encounter corresponds to a unique patient diagnosed with diabetes, although the primary diagnosis may be different.**

During each of the analyzed encounters, lab tests were ordered and medication was administered

Each step was followed by tests for significance of variables with higher degree of freedom, an analysis of deviance table, and sensitivity analysis which was done by removing one variable at the time and looking at changes of beta-coefficients.

the gender variable was not significant in the core model (without HbA1c), it was removed

When tested for sensitivity, the values of beta-coefficients in the model changed by less than 35%, with an exception of the time in the hospital, medical specialty, age, and primary diagnosis that changed by 77%, 47%, 49%, and 65%, respectively, when the discharge disposition was removed. This suggests a relationship between these variables.


# Import libraries

In [16]:
import pandas as pd
pd.options.plotting.backend = "plotly"
import numpy as np

# Load Datasets

In [17]:
data_raw = pd.read_csv("../data/raw/diabetic_data.csv")

# Data Description and Cleaning

Variable identification

List of features and their descriptions in the initial dataset

<table class="table"><tbody><tr><td class="thead-hr" colspan="4"><hr></td></tr><tr class="thead"><td align="left">Feature name</td><td align="center">Type</td><td align="left">Description and values</td><td align="center">% missing</td></tr><tr><td class="thead-hr" colspan="4"><hr></td></tr><tr><td align="left">Encounter ID</td><td align="center">Numeric</td><td align="left">Unique identifier of an encounter</td><td align="center">0%</td></tr><tr><td align="left">Patient number</td><td align="center">Numeric</td><td align="left">Unique identifier of a patient</td><td align="center">0%</td></tr><tr><td align="left">Race</td><td align="center">Nominal</td><td align="left">Values: Caucasian, Asian, African American, Hispanic, and other</td><td align="center">2%</td></tr><tr><td align="left">Gender</td><td align="center">Nominal</td><td align="left">Values: male, female, and unknown/invalid</td><td align="center">0%</td></tr><tr><td align="left">Age</td><td align="center">Nominal</td><td align="left">Grouped in 10-year intervals: <svg height="14.85" id="M2" style="vertical-align:-1.95624pt" version="1.1" viewBox="0 0 5.9875002 14.85" width="5.9875002" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
<g transform="matrix(.017,-0,0,-.017,.062,12.012)"><path d="M290 -163h-170v866h170v-28q-79 -7 -94 -19.5t-15 -72.5v-627q0 -59 14.5 -71.5t94.5 -19.5v-28z" id="x5B"></path></g>
</svg>0, 10), <svg height="14.85" id="M3" style="vertical-align:-1.95624pt" version="1.1" viewBox="0 0 5.9875002 14.85" width="5.9875002" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
<g transform="matrix(.017,-0,0,-.017,.062,12.012)"><path d="M290 -163h-170v866h170v-28q-79 -7 -94 -19.5t-15 -72.5v-627q0 -59 14.5 -71.5t94.5 -19.5v-28z" id="x5B"></path></g>
</svg>10, 20), …, <svg height="14.85" id="M4" style="vertical-align:-1.95624pt" version="1.1" viewBox="0 0 5.9875002 14.85" width="5.9875002" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
<g transform="matrix(.017,-0,0,-.017,.062,12.012)"><path d="M290 -163h-170v866h170v-28q-79 -7 -94 -19.5t-15 -72.5v-627q0 -59 14.5 -71.5t94.5 -19.5v-28z" id="x5B"></path></g>
</svg>90, 100)</td><td align="center">0%</td></tr><tr><td align="left">Weight</td><td align="center">Numeric</td><td align="left">Weight in pounds. </td><td align="center">97%</td></tr><tr><td align="left">Admission type</td><td align="center">Nominal</td><td align="left">Integer identifier corresponding to 9 distinct values, for example, emergency, urgent, elective, newborn, and not available</td><td align="center">0%</td></tr><tr><td align="left">Discharge disposition</td><td align="center">Nominal</td><td align="left">Integer identifier corresponding to 29 distinct values, for example, discharged to home, expired, and not available</td><td align="center">0%</td></tr><tr><td align="left">Admission source</td><td align="center">Nominal</td><td align="left">Integer identifier corresponding to 21 distinct values, for example, physician referral, emergency room, and transfer from a hospital</td><td align="center">0%</td></tr><tr><td align="left">Time in hospital</td><td align="center">Numeric</td><td align="left">Integer number of days between admission and discharge </td><td align="center">0%</td></tr><tr><td align="left">Payer code</td><td align="center">Nominal</td><td align="left">Integer identifier corresponding to 23 distinct values, for example, Blue Cross<i>/</i>Blue Shield, Medicare, and self-pay</td><td align="center">52%</td></tr><tr><td align="left">Medical specialty</td><td align="center">Nominal</td><td align="left">Integer identifier of a specialty of the admitting physician, corresponding to 84 distinct values, for example, cardiology, internal medicine,  family<i>/</i>general practice, and surgeon</td><td align="center">53%</td></tr><tr><td align="left">Number of lab procedures</td><td align="center">Numeric</td><td align="left">Number of lab tests performed during the encounter</td><td align="center">0%</td></tr><tr><td align="left">Number of  procedures</td><td align="center">Numeric</td><td align="left">Number of procedures (other than lab tests) performed during the encounter</td><td align="center">0%</td></tr><tr><td align="left">Number of  medications</td><td align="center">Numeric</td><td align="left">Number of distinct generic names administered during the encounter</td><td align="center">0%</td></tr><tr><td align="left">Number of outpatient visits</td><td align="center">Numeric</td><td align="left">Number of outpatient visits of the patient in the year preceding the encounter</td><td align="center">0%</td></tr><tr><td align="left">Number of emergency visits</td><td align="center">Numeric</td><td align="left">Number of emergency visits of the patient in the year preceding the encounter</td><td align="center">0%</td></tr><tr><td align="left">Number of inpatient visits</td><td align="center">Numeric</td><td align="left">Number of inpatient visits of the patient in the year preceding the encounter</td><td align="center">0%</td></tr><tr><td align="left">Diagnosis 1</td><td align="center">Nominal</td><td align="left">The primary diagnosis (coded as first three digits of ICD9); 848 distinct values</td><td align="center">0%</td></tr><tr><td align="left">Diagnosis 2</td><td align="center">Nominal</td><td align="left">Secondary diagnosis (coded as first three digits of ICD9); 923 distinct values</td><td align="center">0%</td></tr><tr><td align="left">Diagnosis 3</td><td align="center">Nominal</td><td align="left">Additional secondary diagnosis (coded as first three digits of ICD9); 954 distinct values</td><td align="center">1%</td></tr><tr><td align="left">Number of  diagnoses</td><td align="center">Numeric</td><td align="left">Number of diagnoses entered to the system</td><td align="center">0%</td></tr><tr><td align="left">Glucose serum test result </td><td align="center">Nominal</td><td align="left">Indicates the range of the result or if the test was not taken. Values: “&gt;200,” “&gt;300,” “normal,” and “none”  if not measured</td><td align="center">0%</td></tr><tr><td align="left">A1c test result</td><td align="center">Nominal</td><td align="left">Indicates the range of the result or if the test was not taken. Values: “&gt;8” if the result was greater than 8%,  “&gt;7” if the result was greater than 7% but less than 8%, “normal” if the result was less than 7%, and “none” if not measured.</td><td align="center">
									0%</td></tr><tr><td align="left">Change of medications</td><td align="center">Nominal</td><td align="left">Indicates if there was a change in diabetic medications (either dosage or generic name). Values: “change” and “no change”</td><td align="center">0%</td></tr><tr><td align="left">Diabetes medications</td><td align="center">Nominal</td><td align="left">Indicates if there was any diabetic medication prescribed. Values: “yes” and “no”</td><td align="center">0%</td></tr><tr><td align="left">24 features for medications  </td><td align="center">Nominal</td><td align="left">For the generic names: metformin, repaglinide, nateglinide, chlorpropamide, glimepiride, acetohexamide, glipizide, glyburide, tolbutamide, pioglitazone, rosiglitazone, acarbose, miglitol, troglitazone, tolazamide, examide, sitagliptin, insulin, glyburide-metformin, glipizide-metformin, glimepiride-pioglitazone, metformin-rosiglitazone, and metformin-pioglitazone, the feature indicates whether the drug was prescribed or there was a change in the dosage. Values: “up” if the dosage was increased during the encounter, “down” if the dosage was decreased, “steady” if the dosage did not change, and “no” if the drug was not  prescribed</td><td align="center">0%</td></tr><tr><td align="left">Readmitted</td><td align="center">Nominal</td><td align="left">Days to inpatient readmission. Values: “&lt;30” if the patient was readmitted in less than 30 days, “&gt;30” if the patient was readmitted in more than 30 days, and “No” for no record of readmission.</td><td align="center">0%</td></tr><tr class="table-tr"><td colspan="4"><hr class="tbody-hr"></td></tr></tbody></table>


In [18]:
data_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   encounter_id              101766 non-null  int64 
 1   patient_nbr               101766 non-null  int64 
 2   race                      101766 non-null  object
 3   gender                    101766 non-null  object
 4   age                       101766 non-null  object
 5   weight                    101766 non-null  object
 6   admission_type_id         101766 non-null  int64 
 7   discharge_disposition_id  101766 non-null  int64 
 8   admission_source_id       101766 non-null  int64 
 9   time_in_hospital          101766 non-null  int64 
 10  payer_code                101766 non-null  object
 11  medical_specialty         101766 non-null  object
 12  num_lab_procedures        101766 non-null  int64 
 13  num_procedures            101766 non-null  int64 
 14  num_

## Remove duplicated and Not necesary Information

Check if the same patient has more than one encounter.

In [19]:
data_raw["patient_nbr"].duplicated().sum()

30248

remove duplicated patients, First order by readmitted to have `<30` first and then remove duplicated patients, to keep the patiens data with `<30` readmitted.

In [20]:
data_raw = data_raw.sort_values(by='readmitted',ascending= True)

In [21]:
data_raw[['readmitted']].head()

readmitted
83129        <30
28098        <30
86255        <30
28093        <30
28090        <30

In [22]:
data_raw.drop_duplicates(subset ="patient_nbr", keep = "first", inplace = True)

Remove encounter and patient id columns

In [23]:
data_raw.drop(['encounter_id','patient_nbr'],axis='columns',inplace=True)

Number of duplicated data after removing columns

In [24]:
print(f"Duplicated data = {len(data_raw)-len(data_raw.drop_duplicates())}")

Duplicated data = 0


Information of discharge disposition


In our case to find the effectiveness of the treatmento of people how finish the whole hospotal process to reduce the readmission, i'm going to delete the rows related with people that still in hospital or still patient and people who passed away.

- 11,Expired
- 13,Hospice / home
- 14,Hospice / medical facility
- 19,"Expired at home. Medicaid only, hospice."
- 20,"Expired in a medical facility. Medicaid only, hospice."
- 21,"Expired, place unknown. Medicaid only, hospice."

In [25]:
data_raw = data_raw[~data_raw["discharge_disposition_id"].isin([11,13,14,19,20])]

## Handling Missing Values

In [26]:
#look some samples ofthe data
data_raw.sample(10).T

94012      51078      81876    49572  \
race                      AfricanAmerican  Caucasian  Caucasian    Other   
gender                             Female     Female       Male     Male   
age                               [80-90)   [90-100)    [70-80)  [50-60)   
weight                                  ?          ?          ?        ?   
admission_type_id                       1          5          1        1   
discharge_disposition_id                1          1          1        1   
admission_source_id                     7          1          7        7   
time_in_hospital                        2          3          2        2   
payer_code                              ?         MC         MC        ?   
medical_specialty                       ?          ?          ?        ?   
num_lab_procedures                     13         48         41       24   
num_procedures                          0          2          0        3   
num_medications                         9         11         15        9   
number_outpatient                       0          1          0        0   
number_emergency                        0          0          0        0   
number_inpatient                        0          0          0        0   
diag_1                                428        532        428      414   
diag_2                                411        285        413      411   
diag_3                                414        427     250.42      496   
number_diagnoses                        9          9          9        6   
max_glu_serum                        None       None       None     None   
A1Cresult                            None       None       None     None   
metformin                              No         No         No       No   
repaglinide                            No         No         No       No   
nateglinide                            No         No         No       No   
chlorpropamide                         No         No         No       No   
glimepiride                            No         No         No       No   
acetohexamide                          No         No         No       No   
glipizide                              No         No     Steady       No   
glyburide                              Up         No         No       No   
tolbutamide                            No         No         No       No   
pioglitazone                           No         No         No   Steady   
rosiglitazone                          No         No         No       No   
acarbose                               No         No         No       No   
miglitol                               No         No         No       No   
troglitazone                           No         No         No       No   
tolazamide                             No         No         No       No   
examide                                No         No         No       No   
citoglipton                            No         No         No       No   
insulin                                No         No         Up       No   
glyburide-metformin                    No         No         No       No   
glipizide-metformin                    No         No         No       No   
glimepiride-pioglitazone               No         No         No       No   
metformin-rosiglitazone                No         No         No       No   
metformin-pioglitazone                 No         No         No       No   
change                                 Ch         No         Ch       No   
diabetesMed                           Yes         No        Yes      Yes   
readmitted                             NO         NO         NO       NO   

                                     4069       78523      90484  \
race                             Caucasian  Caucasian  Caucasian   
gender                              Female     Female       Male   
age                               [90-100)    [70-80)    [60-70)   
weight                                 

There's some '?' values in the columns, is important to find which values had every column

Get unique values per columns

In [27]:
for col in data_raw:
    print(f"{col} = {data_raw[col].unique()}")

race = ['Caucasian' 'AfricanAmerican' '?' 'Hispanic' 'Other' 'Asian']
gender = ['Female' 'Male' 'Unknown/Invalid']
age = ['[70-80)' '[80-90)' '[10-20)' '[90-100)' '[60-70)' '[50-60)' '[40-50)'
 '[20-30)' '[30-40)' '[0-10)']
weight = ['?' '[50-75)' '[75-100)' '[100-125)' '[125-150)' '[150-175)' '[25-50)'
 '[0-25)' '[175-200)' '>200']
admission_type_id = [1 5 2 3 6 8 4 7]
discharge_disposition_id = [ 6  2  1  3 22  7  5 28  4 25 18  8 15  9 23 12 24 10 17 27 16]
admission_source_id = [ 7  4 17  1  9  6  2  5  3  8 22 20 10 25 13 11 14]
time_in_hospital = [ 3  9  2  8 14  6  4  5  1  7 10 12 13 11]
payer_code = ['MC' 'DM' 'HM' 'SP' '?' 'BC' 'MD' 'CP' 'CM' 'UN' 'CH' 'OG' 'WC' 'MP' 'PO'
 'SI' 'OT' 'FR']
medical_specialty = ['?' 'Psychiatry' 'Surgery-Cardiovascular/Thoracic' 'Nephrology'
 'Podiatry' 'Family/GeneralPractice' 'Gastroenterology' 'InternalMedicine'
 'Emergency/Trauma' 'ObstetricsandGynecology' 'Orthopedics-Reconstructive'
 'Surgery-General' 'Pulmonology' 'Neurology' 'Orthopedics

**Analysis**

- race , weight, payer_code, medical_speciality, diag_2 has '?' value
- gender has 'Unknown/Invalid' value

- This columns of drugs have the same value in all rows:
``` python
examide = ['No']
citoglipton = ['No']
glimepiride-pioglitazone  = ['No']
```
So don't provide any information for the classifier.


In [28]:
# drop examine and cotoglipton columns
data_raw.drop(['examide','citoglipton','glimepiride-pioglitazone'],axis='columns',inplace=True)

Number of 'Unknown/Invalid' value in gender

In [29]:
data_raw["gender"].nunique()

3

There is only 3 rows so i decide to remove the rows with 'Unknown/Invalid' value

In [30]:
data_raw.drop(data_raw[data_raw["gender"] == 'Unknown/Invalid'].index, inplace=True)
print(f"unique gender data = {data_raw['gender'].nunique()}")

unique gender data = 2


### Removing missing values

Change '?' per na value and 

In [31]:
data_raw.replace('?',np.nan,inplace=True)

How Many %NA values per Column

In [32]:
(data_raw.isnull().sum()*100/len(data_raw)).plot.bar(title = 'Percentage of Missing Values per column',color=data_raw.columns)

In [33]:
missing = data_raw.isnull().sum()
missing[missing>0]*100/len(data_raw)

race                  2.711433
weight               96.020754
payer_code           42.982719
medical_specialty    48.086846
diag_1                0.018581
diag_2                0.421651
diag_3                1.710904
dtype: float64


Removing columns `weigth` has more than 95% of the values missing

The `payer_code` column has more than 45% of missing values and 
base on the medical ethics, the company who pays the bill is not related with the service and if is readmitted, consequently  will be removed.

In [34]:
data_raw.drop(['weight','payer_code'],axis='columns',inplace=True)

Analysis of the `medical_speciality`

In [35]:
bar = data_raw["medical_specialty"].value_counts()
bar.plot.bar(title="medical specialty distribution without n.a's", color=bar.index)

I will i'm keeping `medical_specialty` until i perform an statistical analysis of the data,
so the `n.a` values will be replace by `Unknown`

In [36]:
data_raw["medical_specialty"].fillna('Unknow', inplace=True)

Number of duplicated data after removing columns

In [37]:
print(f"Duplicated data = {len(data_raw)-len(data_raw.drop_duplicates())}")

Duplicated data = 0


#### race column analysis

The n.a values are less than the 3% of the data

In [38]:
data_raw["race"].isnull().sum()

1897

How is the distribution

In [39]:
colum = data_raw["race"].unique()
colum

array(['Caucasian', 'AfricanAmerican', nan, 'Hispanic', 'Other', 'Asian'],
      dtype=object)

In [40]:
pd.Series(colum).dropna()

0          Caucasian
1    AfricanAmerican
3           Hispanic
4              Other
5              Asian
dtype: object

In [41]:
bar = data_raw["race"].value_counts()
bar.plot.bar(title="race distribution", color=bar.index)

the majority of people in the dataset are caucasian, therefore i'm going to impute the n.a values with the mode.

In [42]:
data_raw["race"].fillna(data_raw['race'].mode()[0], inplace=True)

In [43]:
data_raw.isnull().sum()

race                           0
gender                         0
age                            0
admission_type_id              0
discharge_disposition_id       0
admission_source_id            0
time_in_hospital               0
medical_specialty              0
num_lab_procedures             0
num_procedures                 0
num_medications                0
number_outpatient              0
number_emergency               0
number_inpatient               0
diag_1                        13
diag_2                       295
diag_3                      1197
number_diagnoses               0
max_glu_serum                  0
A1Cresult                      0
metformin                      0
repaglinide                    0
nateglinide                    0
chlorpropamide                 0
glimepiride                    0
acetohexamide                  0
glipizide                      0
glyburide                      0
tolbutamide                    0
pioglitazone                   0
rosiglitaz

We take care of patients with at least one diagnostic.

In [44]:
data_raw.dropna(subset=["diag_1"], inplace = True)

Check if when `diag_2` is missing `diag_3` too

In [45]:
data_raw["diag_3"].loc[data_raw["diag_2"].isnull()].isnull().sum()

243

if `diag_2` is missing and `diag_3` has a value, copy data from `diag_3` to `diag_2`

because `diag_3` is the aditional secondary diagnosis

In [46]:
data_raw["diag_2"].loc[data_raw["diag_2"].isnull()] = data_raw["diag_3"].loc[data_raw["diag_2"].isnull()].copy()

/tmp/ipykernel_34293/2779560990.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



If the Secondary diagnosis (`diag_2`) is missing will be filled with the primary diagnosis (`diag_1`).

In [47]:
data_raw["diag_2"].loc[data_raw["diag_2"].isnull()] = data_raw["diag_1"].loc[data_raw["diag_2"].isnull()].copy()

/tmp/ipykernel_34293/1485722002.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Finally, If the aditional Secondary diagnosis (`diag_3`) is missing will be filled with the secondary diagnosis (`diag_2`).

In [48]:
data_raw["diag_3"].loc[data_raw["diag_3"].isnull()] = data_raw["diag_2"].loc[data_raw["diag_3"].isnull()].copy()

/tmp/ipykernel_34293/1823848833.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



General Dataframe data

In [49]:
# reset data frame index
data_raw.reset_index(inplace=True, drop=True)

In [50]:
data_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69950 entries, 0 to 69949
Data columns (total 43 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   race                      69950 non-null  object
 1   gender                    69950 non-null  object
 2   age                       69950 non-null  object
 3   admission_type_id         69950 non-null  int64 
 4   discharge_disposition_id  69950 non-null  int64 
 5   admission_source_id       69950 non-null  int64 
 6   time_in_hospital          69950 non-null  int64 
 7   medical_specialty         69950 non-null  object
 8   num_lab_procedures        69950 non-null  int64 
 9   num_procedures            69950 non-null  int64 
 10  num_medications           69950 non-null  int64 
 11  number_outpatient         69950 non-null  int64 
 12  number_emergency          69950 non-null  int64 
 13  number_inpatient          69950 non-null  int64 
 14  diag_1                

In [51]:
# Total Missing values in the dataset

print(f"Total missing values in the dataset = {data_raw.isnull().sum().sum()}")

Total missing values in the dataset = 0


### Save intermediated transformed data
Data without:

 - Duplicates
 - Missings values
 - Non informative data for our taks

In [52]:
# Dataset is small i will be save in a csv file
data_raw.to_csv("../data/interim/diabetes_clean.csv", index = False)

# Partial Results

Initial raw data:
- 101766 rows, 50 Columns , memory usage: 38.8+ MB

After cleaning process
- 69950 rows , 43 Columns , memory usage: 17.5+ MB

# References

- Correction to: Hospital Readmission of Patients with Diabetes - https://link.springer.com/article/10.1007/s11892-018-0989-1

- Center for disease control and prevention, Diabetes atlas- https://gis.cdc.gov/grasp/diabetes/DiabetesAtlas.html

- https://medium.com/@joserzapata/paso-a-paso-en-un-proyecto-machine-learning-bcdd0939d387

- Interpretable ML for Diabetes Patient Readmission - https://www.kaggle.com/aldrinl/interpretable-ml-for-diabetes-patient-readmission

- Prediction on Hospital Readmission - https://www.kaggle.com/iabhishekofficial/prediction-on-hospital-readmission

- Predicting Effective Treatments - https://www.kaggle.com/kavyarall/predicting-effective-treatments/


[Jose R. Zapata](https://joserzapata.github.io)
- https://joserzapata.github.io
- https://twitter.com/joserzapata
- https://www.linkedin.com/in/jose-ricardo-zapata-gonzalez/   